internal/backup_mark.go

```diff
@@ -0,0 +1,81 @@
+package internal
+
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/wal-g/wal-g/internal/storages/storage"
+	"github.com/wal-g/wal-g/internal/tracelog"
+	"github.com/wal-g/wal-g/utility"
+)
+
+// MarkSelfAndPreviousBackupsPermanent marks all previous related backups
+// permanent, including itself, any previous delta backups and initial full
+// backup
+// TODO: unit tests
+func MarkSelfAndPreviousBackupsPermanent(uploader *Uploader, baseBackupFolder storage.Folder, backupName string) error {
```

Comment:
Let's make this function recursive to simplify it.

Reply:
changing the function to recursive will depend on all previous backups to be uncorrupted in order to work, and that may or may not cause some undesirable behavior. On the other hand doing this iteratively limits the number of iterations to `currentSentinel.IncrementCount`, which is not great either but i think can be slightly better. What do you think?

Reply:
If one or more of previous backups is corrupted, then the current backup is useless too. So recursion seems to be OK here.